cmd/jaeger/internal/integration/trace_reader.go

```diff
@@ -0,0 +1,190 @@
+type traceReader struct {
+	logger     *zap.Logger
+	clientConn *grpc.ClientConn
+	client     api_v3.QueryServiceClient
+}
+
+func createTraceReader(logger *zap.Logger, port int) (*traceReader, error) {
+	logger.Info("Creating the trace reader", zap.Int("port", port))
+	opts := []grpc.DialOption{
+		grpc.WithTransportCredentials(insecure.NewCredentials()),
+	}
+
+	cc, err := grpc.NewClient(ports.PortToHostPort(port), opts...)
+	if err != nil {
+		return nil, err
+	}
+
+	return &traceReader{
+		logger:     logger,
+		clientConn: cc,
+		client:     api_v3.NewQueryServiceClient(cc),
+	}, nil
+}
+
+func (r *traceReader) Close() error {
+	r.logger.Info("Closing the gRPC trace reader")
+	return r.clientConn.Close()
+}
+
+func (r *traceReader) GetTraces(
+	ctx context.Context,
+	traceIDs ...tracestore.GetTraceParams,
+) iter.Seq2[[]ptrace.Traces, error] {
+	return func(yield func([]ptrace.Traces, error) bool) {
+		// api_v3 does not support multi-get, so loop through IDs
+		for _, idParams := range traceIDs {
+			idStr := model.TraceIDFromOTEL(idParams.TraceID).String()
+			r.logger.Info("Calling api_v3.GetTrace", zap.String("trace_id", idStr))
+			stream, err := r.client.GetTrace(ctx, &api_v3.GetTraceRequest{
+				TraceId:   idStr,
+				StartTime: idParams.Start,
+				EndTime:   idParams.End,
+			})
+			if !r.consumeTraces(yield, stream, err) {
+				return
+			}
+		}
+	}
+}
+
+func (r *traceReader) GetServices(ctx context.Context) ([]string, error) {
+	res, err := r.client.GetServices(ctx, &api_v3.GetServicesRequest{})
+	if err != nil {
+		return []string{}, err
+	}
+	return res.Services, nil
+}
+
+func (r *traceReader) GetOperations(ctx context.Context, query tracestore.OperationQueryParameters) ([]tracestore.Operation, error) {
+	var operations []tracestore.Operation
+	res, err := r.client.GetOperations(ctx, &api_v3.GetOperationsRequest{
+		Service:  query.ServiceName,
+		SpanKind: query.SpanKind,
+	})
+	if err != nil {
+		return operations, err
+	}
+	for _, operation := range res.Operations {
+		operations = append(operations, tracestore.Operation{
+			Name:     operation.Name,
+			SpanKind: operation.SpanKind,
+		})
+	}
+	return operations, nil
+}
+
+func (r *traceReader) FindTraces(
+	ctx context.Context,
+	query tracestore.TraceQueryParams,
+) iter.Seq2[[]ptrace.Traces, error] {
+	return func(yield func([]ptrace.Traces, error) bool) {
+		if query.NumTraces > math.MaxInt32 {
+			yield(nil, fmt.Errorf("NumTraces must not be greater than %d", math.MaxInt32))
+			return
+		}
+		stream, err := r.client.FindTraces(ctx, &api_v3.FindTracesRequest{
+			Query: &api_v3.TraceQueryParameters{
+				ServiceName:   query.ServiceName,
+				OperationName: query.OperationName,
+				Attributes:    query.Tags,
+				StartTimeMin:  query.StartTimeMin,
+				StartTimeMax:  query.StartTimeMax,
+				DurationMin:   query.DurationMin,
+				DurationMax:   query.DurationMax,
+				//nolint: gosec // G115
+				SearchDepth: int32(query.NumTraces),
+				RawTraces:   true,
```

Comment:
why are we taking the raw traces here? wouldn't it better to let the traces go through the adjusters?

Reply:
because this reader implements the storage API contract, so it needs the raw traces as the normal storage would return.

Reply:
ah got it - thanks for the explanation!